utils/s3/client.cc

```diff
@@ -134,7 +134,7 @@ future<semaphore_units<>> client::claim_memory(size_t size) {
 }
 
 client::group_client::group_client(std::unique_ptr<http::experimental::connection_factory> f, unsigned max_conn)
-        : http(std::move(f), max_conn)
+        : http(std::move(f), max_conn, http::experimental::client::retry_requests::yes)
```

Comment:
What about idempotency? Is retrying a write safe or will it result in double writes? What about deletes? Will they cause spurious failures?

Reply:
Currently we write using multipart upload, so it falls into two -- retry write into part and retry part wrap-up. The latter is metadata manipulation and looks like restarting delete (below). The former restarts just fine, I stepped on it many times when testing Scylla with cstress. The write request carries `partNumber` on board and the data buffers are not moved into socket, but copied. It will be replayed with the same number and AFAICT S3 handles it just fine, returning some valid ETag in response.

Upload wrap-up and delete is harder to say -- both are very short request and I've never seen server closing its connection in the middle of handling those. As long as the request is not finished, it's restartable on the client-side. I can imagine a race when server wraps things up on its side and sends the response, but the connection gets broken by some middle layer after it. Be it multipart-upload-completion or delete, both may result in something like ENOENT.

Reply:
This is why I was skeptical of this retry as a generic Seastar feature. The special "server closed its connection without us noticing" case that the Seastar retry was aimed for is not the only sort of failure that can happen in practice. In rare cases we can see failures in the middle or even right after success of every operation. We can initiate a "delete", S3 may run it, but then we fail to get the response. So we need to do clever retries in client code - e.g., retrying a delete may ignore (?) ENOENT or check if the object was really deleted. I don't see how Seastar can do all of this for us.

So to make a long story short, I think this patch is fine, and improves the retry situation in 99% (or whatever) of the practical cases, but we still need more code to handle the remaining very rare cases. My only worry is that now we'll never see these remaining very rare cases, so we'll forget to handle them. Maybe we need to write a test framework for our S3 client that will inject S3 failures (including broken connections before and after success of the operation).

Reply:
@xemul @avikivity I merged this PR, but please review this comment thread and please think if you should open an issue and/or write a test framework about the kind of failures we can still have that the very specific retry that Seastar now does for us doesn't actually solve.